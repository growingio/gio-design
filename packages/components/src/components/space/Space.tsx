import React, { useContext } from 'react';
import classNames from 'classnames';

import { ConfigContext } from '../config-provider';
import { SpaceProps, SpaceSize } from './interface';

const sizeMap = ['small', 'middle', 'large'];
const size2Number = (size: SpaceSize) => {
  if (typeof size === 'number') {
    return size;
  } else if (sizeMap.indexOf(size) >= 0) {
    return (sizeMap.indexOf(size) + 1) * 8;
  } else {
    return '';
  }
};

const Space: React.FC<SpaceProps> = (props) => {
  const {
    prefixCls: customizePrefixCls,
    className,
    direction = 'horizontal',
    align,
    size = 'small',
    children,
    ...others
  } = props;
  const { getPrefixCls } = useContext(ConfigContext);
  const prefixCls = getPrefixCls('space', customizePrefixCls);
  const cls = classNames(prefixCls, className, {
    [`${prefixCls}-${direction}`]: !!direction,
    [`${prefixCls}-align-${align}`]: !!align,
  });
  const marginDir = direction === 'horizontal' ? 'marginRight' : 'marginBottom';
  const childrenCount = React.Children.count(children);

  if (childrenCount === 0) {
    return null;
  }

  const childrenList = childrenCount > 0 ? children : [children];
  const childNode = React.Children.map(childrenList, (child) => (
    <div className={`${prefixCls}-item`} style={{ [marginDir]: size2Number(size) }}>
      {child}
    </div>
  ));

  return (
    <div className={cls} {...others}>
      {childNode}
    </div>
  );
};

export default Space;
