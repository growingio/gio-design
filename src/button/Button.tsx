import classNames from 'classnames';
import React from 'react';
import { LoadingOutlined } from '@gio-design/icons';
import { ButtonProps } from './interface';
import usePrefixCls from '../utils/hooks/use-prefix-cls';

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
  const {
    type = 'primary',
    size = 'normal',
    loading = false,
    disabled: disabledProp = false,
    htmlType = 'button',
    prefix,
    suffix,
    className,
    children,
    ...restProps
  } = props;

  const prefixCls = usePrefixCls('button');
  const classes = classNames([className, prefixCls], {
    [`${prefixCls}_${type}`]: type,
    [`${prefixCls}_${size}`]: size,
    [`${prefixCls}_loading`]: loading,
  });

  const prefixIcon = loading ? (
    <span className={`${prefixCls}-prefix-icon`}>
      <LoadingOutlined rotating />
    </span>
  ) : (
    prefix && <span className={`${prefixCls}-prefix-icon`}>{prefix}</span>
  );

  const suffixIcon = suffix && <span className={`${prefixCls}-suffix-icon`}>{suffix}</span>;

  const disabled = disabledProp || loading;

  const other: typeof restProps = {
    role: 'button',
    'aria-disabled': disabled,
  };

  return (
    <button
      ref={ref}
      // eslint-disable-next-line react/button-has-type
      type={htmlType}
      className={classes}
      disabled={disabled}
      data-testid="button"
      {...other}
      {...restProps}
    >
      {prefixIcon}
      {children}
      {suffixIcon}
    </button>
  );
});

Button.displayName = 'Button';

Button.defaultProps = {
  type: 'primary',
  size: 'normal',
  loading: false,
  disabled: false,
  htmlType: 'button',
};

export default Button;
