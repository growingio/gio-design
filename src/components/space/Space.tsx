import React,{ useMemo } from 'react';
import classnames from 'classnames';
import toArray from 'rc-util/lib/Children/toArray';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';
import Item from './Item';
import { SpaceProps, SpaceSize } from './interface';
import './style'

export const SpaceContext = React.createContext({
    horizontalSize: 0,
    verticalSize: 0,
    latestIndex: 0,
});

const Space : React.FC<SpaceProps> = props => {
    const {
        size = 8,
        className,
        children,
        direction = 'horizontal',
        prefixCls: customizePrefixCls,
        split,
        style,
        autoWrap = false,
        align = 'center',
    } = props;
    const prefixCls = usePrefixCls('space', customizePrefixCls);
    const [horizontalSize, verticalSize] = useMemo(() => ((Array.isArray(size) ? size : [size, size]) as [SpaceSize, SpaceSize]),[size]);

    const childNodes = toArray(children, { keepEmpty: true });

    if(childNodes.length === 0) {
        return null;
    }

    const itemCls = `${prefixCls}-item`;
    const cls = classnames(
        prefixCls,
        `${prefixCls}-${direction}`,
        `${prefixCls}-align-${align}`,
        className,
    );

    let latestIndex = 0;
    const nodes = childNodes.map((child, index) => {
        if(child !== null && child !== undefined) {
            latestIndex = index;
        }

        return (
          <Item
            className={itemCls}
            // eslint-disable-next-line react/no-array-index-key
            key={index}
            index={index}
            direction={direction}
            split={split}
            autoWrap={autoWrap}
          >
            {child}
          </Item>
        )
    })

    return (
      <div
        className={cls}
        style={{
                ...(autoWrap && { flexWrap: 'wrap', marginBottom: -verticalSize }),
                ...style
            }}
      >
        <SpaceContext.Provider value={{ horizontalSize, verticalSize, latestIndex }}>
          {nodes}
        </SpaceContext.Provider>
      </div>
    )
}

export default Space;