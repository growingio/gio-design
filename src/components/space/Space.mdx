import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Space from './Space';

# Space 间距

<Subtitle>设置组件之间的间距。</Subtitle>

## 参数说明

<ArgsTable of={Space} />

## 代码演示

### 默认

[Example](/?path=/story/basic-components-space--default)
