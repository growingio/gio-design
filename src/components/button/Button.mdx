import { Subtitle, ArgsTable, Story, Canvas } from '@storybook/addon-docs/blocks';
import Button from '.';

# Button 按钮

<Subtitle>点击后即执行相应操作</Subtitle>

## 使用规范

主要按钮 primary

> 突出显示页面上最强烈的动作，应该只在每个屏幕上出现一次(不包括应用程序头或模态对话框)。并不是每个屏幕都需要主按钮。

次要按钮 secondary

> 用于与主按钮配对，以明确有两个选项；或者用于多个无主次之分的操作时。

辅助按钮 assist

> 当页面有许多高频率使用的操作时，将其用作默认按钮。如：看板页面。

文本按钮 text

> 适用于列表内添加多个内容时，如「事件分析」中「添加事件」、「fliter」中「添加过滤条件」等。

图标按钮 icon

> 图标 icon 的颜色可以自定义，图标按钮的存在两种样式：背景为透明度 0%的、浅蓝色背景的。

可扩展

> 按钮具有多种 padding。大多数情况下，我们使用默认紧凑型的 padding，即下列「间距」的规范。还可以将按钮扩展到全宽以填充父容器。

## 参数说明

<ArgsTable of={Button} />

## 代码演示

### 基础

<Canvas>
  <Story id="basic-components-button--default" />
</Canvas>

### 图标

<Canvas>
  <Story id="basic-components-button--icon-button" />
</Canvas>

### 可扩展

<Canvas>
  <Story id="basic-components-button--block-button" />
</Canvas>
