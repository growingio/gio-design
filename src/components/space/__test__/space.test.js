import React from 'react';
import { mount, render } from 'enzyme';
import Space from '..';
import Button from '../../button';

describe('Testing Space.', () => {

    it('should render steady.', () => {
        const wrapper = render(
          <Space>
            Space
            <Button>点击</Button>
            {null}
            {undefined}
            {123}
          </Space>)
        
        expect(wrapper).toMatchSnapshot();
    })

    it('should can define split.', () => {
        const wrapper = mount(
          <Space>
            Space
            {123}
          </Space>
        )
        wrapper.setProps({split: '|'});
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
    })

    it('should render other props.', () => {
        const wrapper = mount(
          <Space>
            Space
            {123}
          </Space>
        )
        wrapper.setProps({size: 8, direction: 'vertical', autoWrap: true});
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
    })

    it('can accept a size array.', () => {
        const wrapper = mount(
          <Space>
            Space
          </Space>
        )
        wrapper.setProps({size: [12, 16]});
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
    })

    it('do not contain anything', () => {
        const wrapper = render(
          <Space />
        )
        expect(wrapper).toMatchSnapshot();
    })
})