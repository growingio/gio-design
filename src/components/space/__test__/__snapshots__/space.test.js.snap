// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Space. can accept a size array. 1`] = `ReactWrapper {}`;

exports[`Testing Space. do not contain anything 1`] = `
Object {
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Testing Space. should can define split. 1`] = `ReactWrapper {}`;

exports[`Testing Space. should render other props. 1`] = `ReactWrapper {}`;

exports[`Testing Space. should render steady. 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "gio-space gio-space-horizontal gio-space-align-center",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "gio-space-item",
          "style": "margin-right:8px",
        },
        "children": Array [
          Object {
            "data": "Space",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "gio-space-item",
            "style": "margin-right:8px",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "gio-btn gio-btn-primary",
                "type": "button",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "data": "点 击",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "span",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "type": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "gio-space-item",
            },
            "children": Array [
              Object {
                "data": "123",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "gio-space-item",
          "style": "margin-right:8px",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "gio-btn gio-btn-primary",
              "type": "button",
            },
            "children": Array [
              Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "data": "点 击",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "span",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "button",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "type": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "type": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "gio-space-item",
          },
          "children": Array [
            Object {
              "data": "123",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "gio-space-item",
            "style": "margin-right:8px",
          },
          "children": Array [
            Object {
              "data": "Space",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "gio-space-item",
        },
        "children": Array [
          Object {
            "data": "123",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "gio-space-item",
            "style": "margin-right:8px",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "gio-btn gio-btn-primary",
                "type": "button",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "data": "点 击",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "span",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "type": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "gio-space-item",
              "style": "margin-right:8px",
            },
            "children": Array [
              Object {
                "data": "Space",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "style": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
